### PR TITLE
Remove Dead Code + Duplication in o.e.c.routing

### PR DESCRIPTION
**Pull request by original-brownbear**
* Removed obviously unused fields+methods
* Inlined public methods that only had one caller
* Simplified `Optional` chain
* Simplified some obviously redundant conditions

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 1]
```diff
@@ -79,8 +79,6 @@
         this.asList = Collections.singletonList(this);
         assert expectedShardSize == UNAVAILABLE_EXPECTED_SHARD_SIZE || state == ShardRoutingState.INITIALIZING ||
             state == ShardRoutingState.RELOCATING : expectedShardSize + " state: " + state;
-        assert expectedShardSize >= 0 || state != ShardRoutingState.INITIALIZING || state != ShardRoutingState.RELOCATING :
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: why remove this?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 2]
```diff
@@ -759,7 +759,7 @@ private void allocateUnassigned() {
             final PriorityComparator secondaryComparator = PriorityComparator.getAllocationComparator(allocation);
             final Comparator<ShardRouting> comparator = (o1, o2) -> {
                 if (o1.primary() ^ o2.primary()) {
-                    return o1.primary() ? -1 : o2.primary() ? 1 : 0;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: why is this ok?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 3]
```diff
@@ -929,7 +929,7 @@ private AllocateUnassignedDecision decideAllocateUnassigned(final ShardRouting s
                             updateMinNode = ((((nodeHigh > repId && minNodeHigh > repId)
                                                    || (nodeHigh < repId && minNodeHigh < repId))
                                                   && (nodeHigh < minNodeHigh))
-                                                 || (nodeHigh > minNodeHigh && nodeHigh > repId && minNodeHigh < repId));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: why is this ok?

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java [hunk 4]
```diff
@@ -143,23 +139,14 @@ public IndexShardRoutingTable shardRoutingTable(ShardId shardId) {
...
-    }
-
     @Nullable
     public ShardRouting getByAllocationId(ShardId shardId, String allocationId) {
-        IndexShardRoutingTable shardRoutingTable = shardRoutingTableOrNull(shardId);
-        if (shardRoutingTable == null) {
-            return null;
-        }
-        return shardRoutingTable.getByAllocationId(allocationId);
+        return Optional.ofNullable(index(shardId.getIndexName()))
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: This is just the code from `shardRoutingTableOrNull` inlined here technically, so there's no change here. In fact, we still `map` exactly twice => we're in the same performance spot as before imo with less code.
That said, I can "de-streamify" this if you want, makes sense to me?

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: This seemed (mathematically it is) already covered by the above `assert`, but I see you point about documentation above. I'll add it back :)

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: `o1.primary() ^ o2.primary() && o1.primary() == false` implies that `o2.primary == true`

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: `nodeHigh > minNodeHigh == true` is implied by the line above `nodeHigh < minNodeHigh == false` (somehow I feel we should maybe indent everything above this line a little differently, it's kinda horrible to read with different levels of precedence at the same indent level imo).

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 5]
```diff
@@ -601,32 +595,35 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
...
-                ShardRouting activeReplica = activeReplicaWithHighestVersion(failedShard.shardId());
-                if (activeReplica == null) {
-                    moveToUnassigned(failedShard, unassignedInfo);
-                } else {
-                    movePrimaryToUnassignedAndDemoteToReplica(failedShard, unassignedInfo);
-                    promoteReplicaToPrimary(activeReplica, routingChangesObserver);
-                }
+                tryPromoteActiveReplica(failedShard, unassignedInfo, routingChangesObserver);
             } else {
-                assert failedShard.primary() == false;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: I guess it seemed somewhat redundant now that this is literally checked 3 lines earlier ? :)

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java [hunk 6]
```diff
@@ -59,11 +59,6 @@
      */
     void relocationSourceRemoved(ShardRouting removedReplicaRelocationSource);
 
-    /**
-     * Called on started primary shard after it has been promoted from replica to primary and is reinitialized due to shadow replicas.
-     */
-    void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java]: Not sure I understand tbh. I guess the logical event described in the JavaDocs above "exists" but this callback is never invoked? Shouldn't we either invoke it when this happens or not have this code?

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java [hunk 7]
```diff
@@ -118,11 +118,7 @@ public IndexShardRoutingTable shardRoutingTable(String index, int shardId) {
         if (indexRouting == null) {
             throw new IndexNotFoundException(index);
         }
-        IndexShardRoutingTable shard = indexRouting.shard(shardId);
-        if (shard == null) {
-            throw new ShardNotFoundException(new ShardId(indexRouting.getIndex(), shardId));
-        }
-        return shard;
+        return OperationRouting.shardRoutingTable(indexRouting, shardId);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: moved it :)

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: I "de-streamified" this now, only 2 more lines anyway :)

**Comment by original-brownbear** [comment]
@ywelsch ping (not urgent obviously :))

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java]: ok, my confusion here. This is obsolete indeed.

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 8]
```diff
@@ -601,32 +595,35 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
...
                 }
             }
-            routingChangesObserver.shardFailed(failedShard, unassignedInfo);
         }
+        routingChangesObserver.shardFailed(failedShard, unassignedInfo);
         assert node(failedShard.currentNodeId()).getByShardId(failedShard.shardId()) == null : "failedShard " + failedShard +
             " was matched but wasn't removed";
     }
 
+    private void tryPromoteActiveReplica(ShardRouting failedShard, UnassignedInfo unassignedInfo,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: perhaps call this `unassignPrimaryAndPromoteActiveReplicaIfExists`? Also add `assert failedShard.primary()`

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java [hunk 9]
```diff
@@ -63,11 +62,19 @@
     // index to IndexRoutingTable map
     private final ImmutableOpenMap<String, IndexRoutingTable> indicesRouting;
 
-    RoutingTable(long version, ImmutableOpenMap<String, IndexRoutingTable> indicesRouting) {
+    private RoutingTable(long version, ImmutableOpenMap<String, IndexRoutingTable> indicesRouting) {
         this.version = version;
         this.indicesRouting = indicesRouting;
     }
 
+    public static IndexShardRoutingTable shardRoutingTable(IndexRoutingTable indexRouting, int shardId) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: add javadocs?

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java [hunk 10]
```diff
@@ -213,42 +209,19 @@ public boolean validate(MetaData metaData) {
...
     }
 
-    public GroupShardsIterator allActiveShardsGrouped(String[] indices, boolean includeEmpty) {
-        return allActiveShardsGrouped(indices, includeEmpty, false);
-    }
-
-    /**
-     * Return GroupShardsIterator where each active shard routing has it's own shard iterator.
-     *
-     * @param includeEmpty             if true, a shard iterator will be added for non-assigned shards as well
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: keep the javadoc?

**Review by ywelsch — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java (hunk 9 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: Done :)

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java (hunk 10 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: Done :)

**Pull request merged**